### PR TITLE
Document the complete list of requirements for the VFS

### PR DESCRIPTION
**Pull request by RaisinTen**
This change attempts to exhaustively document all the requirements of
the VFS. If you disagree with a point or want to add more requirements
to this list, you are more than welcome!

Fixes: https://github.com/nodejs/single-executable/discussions/21
Fixes: https://github.com/nodejs/single-executable/discussions/18
Signed-off-by: Darshan Sen <raisinten@gmail.com>

### CODE REVIEW

File: docs/virtual-file-system-requirements.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+# Optionally support
+
+## Case Sensitivity and Case Preservation
+
+It should be the same as what users of the OS the executable has been packaged
+for expects. This list is based on [case sensitivity in files systems][].
+
+* Unix - case-sensitive
+* MacOS - case-insensitive and case-preserving
+* Windows - case-insensitive and case-preserving
```
Comment (arcanis) [inline docs/virtual-file-system-requirements.md:97]: Why? From our experience w/ zip, forcing case sensitivity within the archives didn't break anything, improved consistency. By constrast, making the code case insensitive would have increased the complexity, worsened the runtime performance, increased the attack surface, for a use case that virtually no-one cares about.

File: docs/virtual-file-system-requirements.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
+Virtual File System Requirements
+================================
+
+This document aims to list all the requirements of the Virtual File System.
+
+# Supported
```
Comment (arcanis) [inline docs/virtual-file-system-requirements.md:6]: I'd add a requirement (perhaps not as a "hard block", but very strongly suggested) that the format should be a good target for third-party software integrations. Specifically:

- We want this format to already have implementation in *multiple* languages (not just JS, since not all tools used in the JS ecosystem are written in JS), ideally production-grade and well-maintained

- We want this format to be consensual enough that third-party tools (VSCode, emacs, ...) won't object to build native integrations with it (for instance, Esbuild recently added zip support to integrate w/ our zip installs; I suspect it'd have been a much harder sell had we use a custom-made format)

File: docs/virtual-file-system-requirements.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+## No interference with valid paths in the file system
+
+If the bundled files in the VFS correspond to certain paths that already exist
+in the real file system, that will be very confusing, so it should use such
+paths that cannot be used by existing files.
+
+Pkg uses [`/snapshot`](https://github.com/vercel/pkg#snapshot-filesystem) as the
+prefix for all the embedded files. This is confusing if `/snapshot` is an
+existing directory on the file system.
```
Comment (cspotcode) [inline docs/virtual-file-system-requirements.md:51]: Docker workflows routinely copy files to, and run things at, the root of the filesystem.

**Comment by cspotcode** [comment]
Adding the list of scenarios to evaluate potential VFS implementations against, copied from #18, as requested here https://github.com/nodejs/single-executable/discussions/18#discussioncomment-3465452

This list will give us a better idea of which modules can be bundled into an SPA as-is, and which use coding techniques that will need to be modified or special-cased to support the SPA runtime.

Static import
static require
dynamic import
dynamic require
passing argv values to FS calls
working directory is `/` (crops up in docker workflows)
being able to manipulate VFS paths as strings vs URL objects
executable is responsible for globbing, for example, it is a test runner

### Globbing scenario

> If I run `./globbing-tool './**/*'`, will it list the contents of the VFS? If the VFS exists at `./globbing-tool` then `./globbing-tool/__vfs__/bin.js` does match the glob.

> If I run `./globbing-tool './**/* | bash-stuff` then `bash-stuff` might try to read `./globbing-tool/__vfs__/bin.js` and fail.

https://github.com/nodejs/single-executable/discussions/18#discussioncomment-3433011

If code within the SPA does naive globbing -- using an off-the-shelf glob library -- then it might end up traversing the VFS. This depends where the VFS is mounted and if `require('fs')` can see the VFS. If `require('fs')` cannot see the VFS, then that causes other problems or limitations which can certainly be documented.

File: docs/virtual-file-system-requirements.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+  // This defaults to the basename of sourceFile, e.g. 'bar' for '/path/bar.js'.
+  namespace?: string;
+```
+
+It might be better to use the single executable path as the base path for the
+files in the VFS, i.e., if the executable has `/a/b/sea` as the path and the VFS
+contains a file named `file.txt`, it would be accessible by the application
+using `/a/b/sea/file.txt`. This approach is similar to how Electron's [ASAR][]
+works, i.e., if the application asar is placed in `/a/b/app.asar`, the
+embedded `file.txt` file would use `/a/b/app.asar/file.txt` as the path.
```
Comment (cspotcode) [inline docs/virtual-file-system-requirements.md:67]: This can be evaluated against potential globbing scenarios where the executable is in the root of the working directory alongside other files to be globbed.  I copied some details from #18 into a comment: https://github.com/nodejs/single-executable/pull/28#issuecomment-1225774154

Comment (arcanis) [reply docs/virtual-file-system-requirements.md:67]: One note regarding deep directory traversal. When implementing a transparent VFS as described here, one thing quickly becomes obvious: you can't make `fs.stat('/path/to/archive.zip')` returns `true` for both `isFile()` and `isDirectory()` (I mean, you can, but it leads to unstable behaviors depending on the order users check them), so you have to pick one or the other.

In Yarn's case we decided that the `fs.stat` result for the archive shouldn't be modified (because it represents an actual physical disk entity), so `isFile()` is true and `isDirectory()` is false (even if, for all other purposes, the archives can be seen as a folder). One advantage of this choice is that deep directory traversals will, by default, not dig into the archives (because they don't see them as directories).

File: README.md [hunk 5]
```diff
@@ -9,6 +9,7 @@ Table of Contents
 -----------------
 
 - [Existing SEA Solutions](./docs/existing-solutions.md)
+- [Virtual File System Requirements](./docs/virtual-file-system-requirements.md)
```
Comment (jviotti) [inline README.md:12]: Maybe worth adding a `- Requirement` item and adding all of these docs under that? i.e.

```
- Requirements
    - [Virtual File System](...)
```

File: docs/virtual-file-system-requirements.md (hunk 2 above)
Comment (jviotti) [reply docs/virtual-file-system-requirements.md:6]: I think both points are very closely related. If a format is very simple to implement, then a lot of implementations will emerge and others projects will not have complains about it.

Maybe the underlying requirement behind this is simplicity?

File: docs/virtual-file-system-requirements.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+## Symbolic links
+
+This is critical for applications that want to use packages like [dugite][] that
+attempt to download [Git executables][] that contain symlinks. Since
+Electron's [ASAR][] does not support symlinks, including [dugite][] as a
+dependency in an Electron app would expand every symlink into individual files,
+thus significantly increase the package size which is not nice.
+
+## Preserve only the executable bit of the file permissions
```
Comment (jviotti) [inline docs/virtual-file-system-requirements.md:22]: I think the word "only" is a bit misleading here. Should it be `Preserve the executable bit` instead?

File: docs/virtual-file-system-requirements.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+It is important to preserve the executable bit of the file permissions, so that
+it is possible for the single-executable to be able to execute only executable
+files. Other than that, all the bundled files would be readable and none will be
+writable.
+
+## Data compression
```
Comment (jviotti) [inline docs/virtual-file-system-requirements.md:29]: I would also rephrase this so say: support different kinds of compressors and compressor settings on individual files. This point was raised before by @ovflowd

File: docs/virtual-file-system-requirements.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+protected by codesigning, making changes to the contents of the VFS should
+invalidate the signature and crash the application if run again. Hence, no write
+operation needs to be supported.
+
+# Optionally support
+
+## Case Sensitivity and Case Preservation
+
+It should be the same as what users of the OS the executable has been packaged
+for expects. This list is based on [case sensitivity in files systems][].
```
Comment (jviotti) [inline docs/virtual-file-system-requirements.md:93]: One potential can of worms here is that case-sensitivity is a characteristic of the underlying file system and not of the OS itself. For example, macOS was case-sensitive for a while, but AFS on newer versions is case-insensitive

File: docs/virtual-file-system-requirements.md (hunk 7 above)
Comment (jviotti) [reply docs/virtual-file-system-requirements.md:29]: Maybe even rephrase this to "Optional Data Compression"? Making sure we allow people to not compress at all is also an important use case for performance reasons

File: docs/virtual-file-system-requirements.md (hunk 2 above)
Comment (ljharb) [reply docs/virtual-file-system-requirements.md:6]: While I agree, literally any format node chooses will automatically become something everyone is willing to integrate - node does have the ability to blaze a trail here.

Comment (arcanis) [reply docs/virtual-file-system-requirements.md:6]: I don't quite agree with that. I can easily imagine projects (especially not-Node-specific ones, like IDEs or file explorers) refusing to spend time integrating a file format supported only by a single runtime, regardless of its popularity.

If the format is a reasonably common one, those concerns are far easier to alleviate - first because they won't have to implement and maintain a parser, and second because this work can be more widely useful (for instance, let's say we pick zip, then those projects essentially get for free support for [wheels archives](https://peps.python.org/pep-0427/), jars, apks, etc).

Additionally, integrations may *already* exist with some file format. For instance, still using zip as an example, both Windows and OSX have native support for exploring their content. Similarly, both emacs and vim can read files from within zip archives. By contrast, I strongly doubt any custom format would end up being supportted by any of these projects.

File: docs/virtual-file-system-requirements.md (hunk 3 above)
Comment (RaisinTen) [reply docs/virtual-file-system-requirements.md:51]: @cspotcode what happens when a file with the same name as the one being copied already exists at the root? I wonder if it makes sense to use such an approach given the discussion regarding addons we are having in https://github.com/nodejs/single-executable/discussions/29#discussioncomment-3474741.

Comment (RaisinTen) [reply docs/virtual-file-system-requirements.md:51]: Ah, I initially understood that you thought the `/snapshot` approach was a good idea and that's why you shared the docker example, my bad :)

Comment (cspotcode) [reply docs/virtual-file-system-requirements.md:51]: Ok, that's good to hear. :)  Agreed: I think we should *not* use `/snapshot` or any prefix at the root of the filesystem, because it may conflict with files on the real filesystem.  More specifically, a user of a node SPA may ask that SPA to read and/or write files within the `/snapshot` directory.  Choosing a different name for the VFS may reduce or obscure this possibility, but does not eliminate it.

Example pseudocode of a `Dockerfile`:

```dockerfile
# Suppose that build-tool is a node SPA
# but the consumers of build-tool do not know and should not need to know that it is a node SPA
FROM ubuntu
COPY ./build-tool /build-tool
COPY ./sources /sources
# This invocation of build-tool expects it to read from and write to files within the `/snapshot` directory
# on the _real_ filesystem
RUN /build-tool /sources/index.foo /snapshot && rm -rf /sources /build-tool
ENTRYPOINT /snapshot/shim.sh
```

**Review by ovflowd — COMMENTED** [review]
There's just one section I feel it is missing... Something regarding ensuring that the VFS respects the paths defined within the application.

What I mean by this is that by having a piece of code that wants to access a file from the VFS (e.g., A simple JSON file), when SEA is building the binary, it needs to ensure that these paths are transformed correctly.

Not sure if this is covered on "No interference with valid paths in the file system", but I want to ensure that if the scenario where a file exists both in the VFS and real-file-system for the same path (e.g.: `./my-file.txt` and if the current working directory also has a `my-file.txt`) that the VFS is smart enough to distinguish what should be accessed.

Another essential point that I feel is not covered here are FS permissions. Like, how we deal with scenarios where we don't have access to the actual file system or how we deal with insufficient permissions...

Finally, something mentioned [here](https://github.com/nodejs/single-executable/discussions/29) is how we deal with large VFSs as we don't want to deal with the application being slow on its bootstrap and execution because of random file accesses and the need to unarchive the whole VFS...

**Comment by ovflowd** [comment]
BTW, may be worth adding a benchmarking sub-section to the "supported" area, as we should investigate the performance under numerous circumstances and file systems and architectures of the performance and reliability of some of the VFSs we want to support. (Either the ones we're creating, or the ones we think about supporting)

**Comment by cspotcode** [comment]
> Not sure if this is covered on "No interference with valid paths in the file system"

I believe this is what we're thinking about in [arcanis's explanation](https://github.com/nodejs/single-executable/pull/28#discussion_r953867475) for how they mount zip files and avoid accidental deep directory traversal, and the [thread](https://github.com/nodejs/single-executable/pull/28#discussion_r953841581) about why mounting at `/snapshot` is a bad idea.

We could break "no interference" into multiple sub-requirements?  Pick scenarios from the list here https://github.com/nodejs/single-executable/pull/28#issuecomment-1225774154 appended with both "on the real filesystem" and "on the VFS"

For example:
dynamic require can execute a file on the VFS
dynamic require can execute a file on the real filesystem
`readFileSync` can read any file on the real filesystem
`readFileSync` can read any file on the VFS

> when SEA is building the binary, it needs to ensure that these paths are transformed correctly.

To me, this implies that all VFS paths are known at build-time and somehow transformed.  So dynamic `readFileSync`, dynamic import, dynamic require might fail at runtime.  I've been operating under the assumption that we're _not_ talking about a build-time transformation.  If other people are assuming a build-time transformation, e.g. somehow rewriting source-code to transform `readFile('./foo')` into `readFile(new URL('nodevfs://foo))`, then we can assign names to the two ideas to make them easier to talk about.

**Comment by cspotcode** [comment]
In the spirit of https://github.com/nodejs/single-executable/pull/28#issuecomment-1228842682:

Should we also add a doc that explains how each VFS works internally?  E.g. a primer explaining the characteristics of a Zip file: table of contents at the end for random access, compression is optional but supported, compression can be enabled per-file, can store symlinks.

Then the same for the other VFS implementations being considered.

Not sure if this already exists and I missed it, or if it should be a separate PR.

**Comment by ljharb** [comment]
They don’t have to be known at build-time - the fs wrapper can provide privileged access inside the SEA to dynamic paths.

**Comment by ovflowd** [comment]
> To me, this implies that all VFS paths are known at build-time and somehow transformed. So dynamic readFileSync, dynamic import, dynamic require might fail at runtime. I've been operating under the assumption that we're not talking about a build-time transformation. If other people are assuming a build-time transformation, e.g. somehow rewriting source-code to transform readFile('./foo') into readFile(new URL('nodevfs://foo)), then we can assign names to the two ideas to make them easier to talk about.

I don't think they don't need to be known at build-time, I was just thinking something like a Proxy or something that transparently wraps around `fs` so it operates during runtime resolution. But to be honest, it would be neat to transform static imports.

**Comment by ovflowd** [comment]
> Should we also add a doc that explains how each VFS works internally? E.g. a primer explaining the characteristics of a Zip file: table of contents at the end for random access, compression is optional but supported, compression can be enabled per-file, can store symlinks.

+1

**Comment by jviotti** [comment]
> Another essential point that I feel is not covered here are FS permissions. Like, how we deal with scenarios where we don't have access to the actual file system or how we deal with insufficient permissions...

Permissions are usually tied to the group/owner, etc distinction that is specific to one instance of the file system and won't translate to other file systems with a different user setup. 

That said, I think there is a subset of permissions that can be preserved. Execution is the main one that comes to mind. Tools like ASAR omit all permissions but the executable bit for this reason. The other aspect would be whether a file is readable or writable, but based on various threads so far, I think we agreed that the VFS is read-only, so we can probably ignore this characteristic.

> BTW, may be worth adding a benchmarking sub-section to the "supported" area, as we should investigate the performance under numerous circumstances and file systems and architectures of the performance and reliability of some of the VFSs we want to support. (Either the ones we're creating, or the ones we think about supporting)

Yeah, this is exactly what I'm hoping for. I wanted to collect all the requirements we want first (without being biased too much by existing solutions), and then use that as the basis for checking which VFS out there, if any, checks the boxes or not.

> If other people are assuming a build-time transformation, e.g. somehow rewriting source-code to transform readFile('./foo') into readFile(new URL('nodevfs://foo)), then we can assign names to the two ideas to make them easier to talk about.

This sounds very hard to do right, mainly with dynamic requires (which we should support). I didn't study this aspect much, but PKG seems to do it very well, so we should at least try to copy that approach cc @jesec 

> Should we also add a doc that explains how each VFS works internally? E.g. a primer explaining the characteristics of a Zip file: table of contents at the end for random access, compression is optional but supported, compression can be enabled per-file, can store symlinks.

That would be amazing. Can you open a GitHub Issue for this?

File: docs/virtual-file-system-requirements.md (hunk 2 above)
Comment (RaisinTen) [reply docs/virtual-file-system-requirements.md:6]: I've added this to the doc

File: docs/virtual-file-system-requirements.md (hunk 1 above)
Comment (RaisinTen) [reply docs/virtual-file-system-requirements.md]: Agreed, updated the doc to make the VFS case-sensitive.

File: docs/virtual-file-system-requirements.md (hunk 3 above)
Comment (RaisinTen) [reply docs/virtual-file-system-requirements.md]: Thanks, added the docker example

File: docs/virtual-file-system-requirements.md (hunk 4 above)
Comment (RaisinTen) [reply docs/virtual-file-system-requirements.md:69]: I've added the stat info under globbing but I'm open to flipping the return values of `isDirectory()` and `isFile()` if y'all have a preference.

File: README.md (hunk 5 above)
Comment (RaisinTen) [reply README.md]: done

File: docs/virtual-file-system-requirements.md (hunk 7 above)
Comment (RaisinTen) [reply docs/virtual-file-system-requirements.md]: done

File: docs/virtual-file-system-requirements.md (hunk 8 above)
Comment (RaisinTen) [reply docs/virtual-file-system-requirements.md]: Very interesting, documented the VFS to be case-sensitive now. It's less confusing.

**Review by jviotti — APPROVED** [review]
Looks good. We can merge this and send further PRs to clarify later

**Comment by RaisinTen** [comment]
@cspotcode 

> Static import
static require
dynamic import
dynamic require

I have added a section about this to the docs but I'm not sure what problems we might face while implementing that. Might be better if you could share that info.

> passing argv values to FS calls

Could you expand on that?

> working directory is / (crops up in docker workflows)

I believe that one is already covered in the "No interference with valid paths in the file system" section.

> being able to manipulate VFS paths as strings vs URL objects

Added that in but would like to know if there are any gotchas.

> executable is responsible for globbing, for example, it is a test runner

Added a "Globbing" section.


@ovflowd 

> Another essential point that I feel is not covered here are FS permissions. Like, how we deal with scenarios where we don't have access to the actual file system or how we deal with insufficient permissions...

Are you referring to the implications our VFS might have when https://github.com/nodejs/node/pull/44004 lands?

**Comment by ovflowd** [comment]
> Are you referring to the implications our VFS might have when https://github.com/nodejs/node/pull/44004 lands?

Yea. 😅 (That PR is def a big break-change).

**Review by ovflowd — APPROVED** [review]
:shipit:

**Comment by cspotcode** [comment]
@RaisinTen

Thanks, I meant the list as a checklist to evaluate potential VFS implementations.  Some are not problems with the current VFS proposals but might be problems with future VFS proposals.

> dynamic import
> dynamic require

I suppose `dynamic` is more important than `static`.  If an SPA does `require(require.resolve('./tool.config.js'))` and wants to execute a file from the _real_ filesystem, it should be able to.  And if it tries to `require(require.resolve('bundled-internal-dependency'))` to execute something from the VFS, then that should also work.  Mostly wanted to include this as a reason *not* to rely on build-time resolution the way that bundlers do.

> being able to manipulate VFS paths as strings vs URL objects

If someone proposes that the VFS exist at a `vfs-file://` prefix, then this might become an issue.  fs APIs accept URL objects, but this means code in (transitive) dependencies which assumes all native paths are strings may fail when passed URL objects.  Perhaps a (transitive) dependency uses `require.resolve`.  I'm not sure if `require.resolve()` is meant to work with URLs.

I can imagine someone proposing `vfs-file://` as a solution for "No interference with valid paths in the file system."

> passing argv values to FS calls

We can have a couple examples of (contrived) corner-cases such as `./bundled-spa-formatter fmt ./bundled-spa/snapshot/index.js`.
Internally, I imagine the formatter will probably do something akin to `fs.readFileSync(process.argv[3])`, format it, and echo the results to stdout.  In this example, `process.argv[3]` is a path in the VFS.  I don't think this is a bad thing -- the example is contrived -- but it might be worth documenting as a well-known quirk of SPAs.

> working directory is / (crops up in docker workflows)

Thanks, yeah looks like it's covered.

File: docs/virtual-file-system-requirements.md [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+## Preserve file-hierarchy information
+
+A filesystem is incomplete without this because there's no way for the
+single-executable to be able to access nested file paths.
+
+## No interference with valid paths in the file system
+
+If the bundled files in the VFS correspond to certain paths that already exist
+in the real file system, that will be very confusing, so it should use such
+paths that cannot be used by existing files.
```
Comment (cspotcode) [inline docs/virtual-file-system-requirements.md:47]: More than being confusing, this would outright break certain use-cases.

**Comment by jviotti** [comment]
> If someone proposes that the VFS exist at a vfs-file:// prefix, then this might become an issue.

I think we all agree that we can officially drop that idea, as introducing a prefix would break a huge amount of things.

> If an SPA does require(require.resolve('./tool.config.js')) and wants to execute a file from the real filesystem, it should be able to. And if it tries to require(require.resolve('bundled-internal-dependency')) to execute something from the VFS, then that should also work.

I think this is the best summary of the entire problem so far. We need a clear answer to this problem. The scenarios that we need to handle are:

1. The given file path exists in both the VFS and the outer file system
2. The given file path exists in the VFS but not in the outer file system
3. The given file path does not exist in the VFS but exists in the outer file system
4. The given file path doesn't exist anywhere

All the cases except from the first one are trivial. For the first one, a potentially good-enough way of solving the dilemma is a precedence rule. We i.e. prefer the VFS location over the outer file system or vice-versa.

However, which one we prefer is still a tricky problem, and I can see how it would give us pros and cons both ways.

What do you think? Is there an approach other than precedence we would consider, that does not involve inventing an arbitrary prefix for the VFS?

**Comment by cspotcode** [comment]
I think we should do what yarn does for zips: the VFS is available on sub-paths of the executable, as if the executable itself were a directory even though it is not.

For example, if the executable is at `/foo/packaged-cli-tool` then `fs.readFileSync('/foo/packaged-cli-tool/nested/path/into/the/vfs/index.js')` will work.

**Comment by jviotti** [comment]
Ah, that's brilliant. I love it. I can't think of any drawback to that. @RaisinTen shall we document that approach?

**Comment by RaisinTen** [comment]
> @RaisinTen shall we document that approach?

@jviotti does this look okay https://github.com/nodejs/single-executable/blob/b7f5638bc36c441f6721d4832ca89382b357c405/docs/virtual-file-system-requirements.md?plain=1#L64-L69?

**Comment by jviotti** [comment]
@RaisinTen Looks good. My only comment is that rather than saying "It might be better...", say "A possible solution...", etc

File: docs/virtual-file-system-requirements.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+`fs.statSync(process.execPath).isFile()` will return `true`. That way, if code
+within the single-executable does naive globbing using an off-the-shelf glob
+library, VFS file paths, which are considered internal to the single-executable,
+would not interfere with the actual file paths.
+
+## Accept file paths in the VFS as arguments
+
+If a single-executable formatter is run with an argument that is a path to a
+file inside the VFS, it should be able to use the `fs` APIs to read, format and
+print the formatted contents to `stdout`.
```
[the remark this answers is not in the document]
Comment (jviotti) [reply docs/virtual-file-system-requirements.md:83]: I guess the decision we need to make here is: can files within the VFS be referenced from outside the VFS, or only internally?

If we have an approach that disambiguates files inside and outside of the VFS (like the base path approach), then I don't see why we should not allow this.

[the remark this answers is not in the document]
Comment (jviotti) [reply docs/virtual-file-system-requirements.md:83]: For those fundamental questions, let's try to document them in the README. I think it will be useful to have a concrete trail of fundamental questions & answers somewhere.

[the remark this answers is not in the document]
Comment (jviotti) [reply docs/virtual-file-system-requirements.md:83]: I'll start GitHub Discussions on a separate category to track these down, and start a PR documenting the results.

[the remark this answers is not in the document]
Comment (RaisinTen) [reply docs/virtual-file-system-requirements.md:82]: I guess most people in https://github.com/nodejs/single-executable/discussions/18 were in favor of keeping the paths inside the VFS transparent? I have also changed the globbing part to better reflect that.

**Comment by RaisinTen** [comment]
@jviotti done, PTAL

**Comment by RaisinTen** [comment]
@addaleax I have also moved the data compression part to the optional section since you mentioned that during the meeting, PTAL.

**Comment by jviotti** [comment]
@RaisinTen Let's merge it. If there are further comments, we can send further PRs :)
